migrations/20210612_initial.sql

```diff
@@ -0,0 +1,104 @@
+CREATE TABLE profiles (
+  profile_id INTEGER NOT NULL PRIMARY KEY,
+  display_name TEXT NOT NULL,
+  properties TEXT NOT NULL -- JSON with user or contact profile
+);
+
+CREATE TABLE known_servers(
+  host TEXT NOT NULL,
+  port TEXT NOT NULL,
+  key_hash BLOB,
+  PRIMARY KEY (host, port)
+) WITHOUT ROWID;
+
+CREATE TABLE contacts (
+  contact_id INTEGER NOT NULL PRIMARY KEY,
+  display_name TEXT NOT NULL, -- display name set by local user
+  properties TEXT NOT NULL, -- JSON set by local user
+  profile_id INTEGER REFERENCES profiles, -- sent by remote contact
+  contact_status TEXT NOT NULL DEFAULT ''
+);
+
+CREATE TABLE contact_connections (
+  connection_id BLOB NOT NULL PRIMARY KEY,
+  contact_id INTEGER NOT NULL REFERENCES contacts ON DELETE RESTRICT,
+  conn_level INTEGER NOT NULL DEFAULT 0,
+  via_conn BLOB REFERENCES contact_connections (connection_id),
+  conn_status TEXT NOT NULL DEFAULT ''
+) WITHOUT ROWID;
+
+CREATE TABLE contact_invitations (
+  invitation_id BLOB NOT NULL PRIMARY KEY,
+  contact_id INTEGER NOT NULL REFERENCES contacts ON DELETE RESTRICT,
+  invitation_status TEXT NOT NULL DEFAULT ''
+) WITHOUT ROWID;
+
+CREATE TABLE groups (
+  group_id INTEGER NOT NULL PRIMARY KEY, -- local group ID
+  display_name TEXT NOT NULL, -- shared
+  properties TEXT NOT NULL -- shared JSON group profile
+);
+
+CREATE TABLE group_members ( -- group members, nullable fields are NULL for the local user
+  group_id INTEGER NOT NULL REFERENCES groups ON DELETE RESTRICT,
+  member_id BLOB NOT NULL, -- shared member ID, unique per group
+  member_role TEXT NOT NULL DEFAULT '', -- owner, admin, moderator, ''
+  member_status TEXT NOT NULL DEFAULT '', -- inv | con | full | off
+  contact_id INTEGER REFERENCES contacts ON DELETE RESTRICT,
+  invited_by INTEGER REFERENCES contacts ON DELETE RESTRICT,
+  connection_id BLOB UNIQUE,
+  PRIMARY KEY (group_id, member_id)
+) WITHOUT ROWID;
+
+CREATE TABLE events ( -- messages received by the agent, append only
+  event_id INTEGER NOT NULL PRIMARY KEY,
```

Comment:
switch to manually reserved, incremented and set id? we'll have the same problem retrieving this id as we had for messages. particularly it's not reliable for concurrent use as far as I remember

Reply:
wait... I thought it's ok to auto increment if it's in the same transaction?

Reply:
https://sqlite.org/c3ref/last_insert_rowid.html

in transaction it should be fine probably, it's all shady in their documentation

```
For the purposes of this routine, an INSERT is considered to be successful even if it is subsequently rolled back.
...
If a separate thread performs a new INSERT on the same database connection while the sqlite3_last_insert_rowid() function is running and thus changes the last insert rowid, then the value returned by sqlite3_last_insert_rowid() is unpredictable and might not equal either the old or the new last insert rowid.
```

I guess we wouldn't have such problems, but I don't like relying on it. In agent I believe it allowed us to break a transaction